packages/web/src/store/application/ui/buy-audio/sagas.ts

```diff
@@ -0,0 +1,460 @@
+import { TransactionHandler } from '@audius/sdk/dist/core'
+import { Jupiter, SwapMode, RouteInfo } from '@jup-ag/core'
+import { AccountInfo, Token, TOKEN_PROGRAM_ID } from '@solana/spl-token'
+import {
+  Connection,
+  Keypair,
+  LAMPORTS_PER_SOL,
+  PublicKey,
+  Transaction
+} from '@solana/web3.js'
+import BN from 'bn.js'
+import JSBI from 'jsbi'
+import {
+  call,
+  select,
+  put,
+  takeEvery,
+  race,
+  delay
+} from 'typed-redux-saga/macro'
+
+import { BNWei } from 'common/models/Wallet'
+import { getAccountUser } from 'common/store/account/selectors'
+import { TOKEN_LISTING_MAP } from 'common/store/buy-audio/constants'
+import {
+  exchange,
+  exchangeAfterBalanceChange,
+  exchangeFailed,
+  exchangeSucceeded,
+  quote,
+  quoteFailed,
+  quoteSucceeded
+} from 'common/store/buy-audio/slice'
+import { increaseBalance } from 'common/store/wallet/slice'
+import { convertJSBIToUiString, weiToString } from 'common/utils/wallet'
+import AudiusBackend from 'services/AudiusBackend'
+import { waitForLibsInit } from 'services/audius-backend/eagerLoadUtils'
+
+const BALANCE_CHANGE_TIMEOUT_MS = 120_000 // 2 minutes
+const BALANCE_CHANGE_POLL_INTERVAL_MS = 5_000 // 5 second
+const WEI_DECIMALS = 18 // 18 decimals on ETH AUDIO
+const SPL_DECIMALS = 8 // 8 decimals on SPL AUDIO
+const SWAP_MIN_SOL = 0.05 // Minimum SOL balance needed to execute a Jupiter Swap
+const ERROR_CODE_SLIPPAGE = 6000 // Error code for when the swap fails due to specified slippage being exceeded
+let _jup: Jupiter
+
+const rpcList = [
+  'https://solana-api.projectserum.com',
+  'https://ssc-dao.genesysgo.net',
+  'https://mercuria-fronten-1cd8.mainnet.rpcpool.com'
+] // taken from jup.ag main site
+
+function convertWAudioToWei(amount: BN) {
```

Comment:
surprised this doesn't exist somewhere? maybe put in a util file like utils/wallet?